hedera-node/docs/design/services/token-service/airdrops/airdrop-pending-state.md

```diff
@@ -0,0 +1,126 @@
+
+`StateDefinition.*onDisk*(*AIRDROPS_KEY*, PendingAirdropId.PROTOBUF, PendingAirdropValue.PROTOBUF, MAX__AIRDROPS);`
+
+We can have a case, where we have multiple airdrops for the same fungible token. In this case the  `PendingAirdropValue` would be updated with the new aggregated value.
+
+### Serialization
+
+In order for the pending state to be kept during node upgrades or restarts we should write it on disk with a serialization mechanism, as we do for other context or state related information. This could be done in the `InitialModServiceTokenSchema` since the pending airdrop state would be written in the already existing token state.
+
+We should define a new entry in `StateChildIndices` called `AIRDROPS`. Based on it, we would fetch a `VirtualMap` entry for the airdrops. It would include an `AirdropKey`, which would be a `PendingAirdropId` representation as a key and an `EntityNum` , which would be a `PendingAirdropValue` representation as a value. The `AirdropKey` should implement the `VirtualKey` interface, so that the key has a very strong hash algorithm to optimize performance. `EntityNum` is an already existing numeric wrapper for long values that extend the `VirtualValue` interface and provide better efficiency for usage in `VirtualMaps`.
+
+We would also need to create genesis children for this map in `ServicesState`  and a boolean property for enabling/disabling virtual merkle state from them - `tokens.airdrops.useVirtualMerkle` .
+
+### Stores
+
+We need to define a special `WritableStore` for the airdrops, so that we can manipulate the content in them. This would be the `WritableAirdropStore`.
+
+It would extend `ReadableAirdropStoreImpl` that implements the following interface:
+
+```java
+public interface ReadableAirdropStore {
+	PendingAirdropValue get(@NonNull final PendingAirdropId airdropId);
+	long sizeOfState();
+	void warm(@NonNull final PendingAirdropId airdropId);
+}
+```
+
+The `ReadableAirdropStoreImpl` would keep an instance of `ReadableKVState<PendingAirdropId, PendingAirdropValue> airdropsState`.
+
+The `WritableAirdropStore` would add additional methods for operating over the airdrops state, so it would have the following methods:
+
+```java
+public class WritableAirdropStore {
+
+    void put(@NonNull final PendingAirdropId airdropId, @NonNull final PendingAirdropValue airdropValue) {
+        // invoked by the `TokenAirdropTransaction`
+    }
+
+    void remove(@NonNull PendingAirdropId airdropId) {
+        // invoked by the `TokenCancelAirdrop`
+    }
+}
+```
+
+### Operations
+
+1. Fetching airdrop values from the collections. This would be applicable only for fungible airdrops.
```

Comment:
Q: wouldn't this also be applicable when Claiming a fungible airdrops. We should call out that the get will be done in 2 transaction scenarios
```suggestion
1. Fetching airdrop values from the collections. This would be applicable only for fungible tokens (both TokenAirdrop and TokenClaimAirdrop transaction).
```